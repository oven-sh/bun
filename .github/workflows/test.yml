name: Test

concurrency:
  group: ${{ github.workflow }}-${{ inputs.run-id || github.run_id }}
  cancel-in-progress: true

on:
  push:
  workflow_call:
    inputs:
      run-id:
        type: string
        required: true

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/download-artifact@v4
        with:
          run-id: ${{ inputs.run-id || '8623134412' }}
          pattern: bun-*
      - run: |
          ls -la
