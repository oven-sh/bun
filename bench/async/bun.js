import { bench, run } from "mitata";

bench("sync", () => {});
bench("async", async () => {});
bench("await 1", async () => await 1);

await run();
