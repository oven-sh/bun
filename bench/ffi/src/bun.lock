{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@napi-rs/cli": "^2.10.1",
        "@node-rs/helper": "^1.3.3",
      },
    },
  },
  "packages": {
    "@napi-rs/cli": ["@napi-rs/cli@2.14.8", "", { "bin": { "napi": "scripts/index.js" } }, "sha512-IvA3s8BqohMdUbOkFn7+23u1dhIJZCkA8Xps7DD4SLdCMbcbUF6MUuKiqxuqmVHBFTaxU25sU56WdX3efqGgPw=="],

    "@napi-rs/triples": ["@napi-rs/triples@1.1.0", "", {}, "sha512-XQr74QaLeMiqhStEhLn1im9EOMnkypp7MZOwQhGzqp2Weu5eQJbpPxWxixxlYRKWPOmJjsk6qYfYH9kq43yc2w=="],

    "@node-rs/helper": ["@node-rs/helper@1.3.3", "", { "dependencies": { "@napi-rs/triples": "^1.1.0" } }, "sha512-p4OdfQObGN9YFy5WZaGwlPYICQSe7xZYyXB0sxREmvj1HzGKp5bPg2PlfgfMZEfnjIA882B9ZrnagYzZihIwjA=="],
  }
}
