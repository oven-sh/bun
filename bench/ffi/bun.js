import { ptr, dlopen, CString, toBuffer } from "bun:ffi";
import { run, bench, group } from "mitata";

const { napiNoop, napiHash, napiString } = require(import.meta.dir + "/src/ffi_napi_bench.node");

const {
  symbols: {
    ffi_noop: { native: ffi_noop },
    ffi_hash: { native: ffi_hash },
    ffi_string: { native: ffi_string },
  },
} = dlopen(import.meta.dir + "/src/ffi_napi_bench.node", {
  ffi_noop: { args: [], returns: "void" },
  ffi_string: { args: [], returns: "ptr" },
  ffi_hash: { args: ["ptr", "u32"], returns: "u32" },
});

const bytes = new Uint8Array(64);

group("bun:ffi", () => {
  bench("noop", () => ffi_noop());
  bench("hash", () => ffi_hash(ptr(bytes), bytes.byteLength));

  bench("c string", () => new CString(ffi_string()));
});

if (process.env.SHOW_NAPI)
  group("bun:napi", () => {
    bench("noop", () => napiNoop());
    bench("hash", () => napiHash(bytes));

    bench("string", () => napiString());
  });

await run();
