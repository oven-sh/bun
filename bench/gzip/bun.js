import { run, bench, group } from "mitata";
import { gzipSync, gunzipSync } from "bun";

const data = new TextEncoder().encode("Hello World!".repeat(9999));

const compressed = gzipSync(data);

const libraries = ["zlib"];
if (Bun.semver.satisfies(Bun.version, ">=1.1.21")) {
  libraries.push("libdeflate");
}

const options = { library: undefined };
const benchFn = (name, fn) => {
  if (libraries.length > 1) {
    group(name, () => {
      for (const library of libraries) {
        bench(library, () => {
          options.library = library;
          fn();
        });
      }
    });
  } else {
    bench(name, () => {
      fn();
    });
  }
};

benchFn(`roundtrip - "Hello World!".repeat(9999))`, () => {
  gunzipSync(gzipSync(data, options), options);
});

benchFn(`gzipSync("Hello World!".repeat(9999)))`, () => {
  gzipSync(data, options);
});

benchFn(`gunzipSync("Hello World!".repeat(9999)))`, () => {
  gunzipSync(compressed, options);
});

await run();
