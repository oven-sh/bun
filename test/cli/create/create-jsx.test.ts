import "bun";
import { expect, test, describe, beforeEach, afterAll } from "bun:test";
import { tempDirWithFiles as tempDir, bunExe, bunEnv, isCI, isWindows } from "harness";
import { cp, readdir } from "fs/promises";
import path from "path";
import puppeteer, { type Browser } from "puppeteer";
import type { Subprocess } from "bun";
import * as vm from "vm";
const env = {
  ...bunEnv,
};
const baseOptions = {
  dumpio: !!process.env.CI_DEBUG,

  args: [
    "--disable-gpu",
    "--disable-dev-shm-usage",
    "--disable-setuid-sandbox",
    "--no-sandbox",
    "--ignore-certificate-errors",
    "--use-fake-ui-for-media-stream",
    "--use-fake-device-for-media-stream",
    "--disable-sync",
  ],
  executablePath: process.env.BROWSER_EXECUTABLE,
  headless: true,
};
let puppeteerBrowser: Browser | null = null;
async function getPuppeteerBrowser() {
  if (!puppeteerBrowser) {
    puppeteerBrowser = await puppeteer.launch(baseOptions);
  }
  return puppeteerBrowser;
}

afterAll(async () => {
  if (puppeteerBrowser) {
    await puppeteerBrowser.close();
  }
});

async function getServerUrl(process: Subprocess<any, "pipe", any>, all = { text: "" }) {
  // Read the port number from stdout
  const decoder = new TextDecoder();
  let serverUrl = "";
  all.text = "";

  const reader = process.stdout.getReader();
  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    const textChunk = decoder.decode(value, { stream: true });
    all.text += textChunk;
    console.log(textChunk);

    if (all.text.includes("http://")) {
      serverUrl = all.text.trim();
      serverUrl = serverUrl.slice(serverUrl.indexOf("http://"));

      serverUrl = serverUrl.slice(0, serverUrl.indexOf("\n"));
      if (URL.canParse(serverUrl)) {
        break;
      }

      serverUrl = serverUrl.slice(0, serverUrl.indexOf("/n"));
      serverUrl = serverUrl.slice(0, serverUrl.lastIndexOf("/"));
      serverUrl = serverUrl.trim();

      if (URL.canParse(serverUrl)) {
        break;
      }
    }
  }
  reader.releaseLock();

  if (!serverUrl) {
    throw new Error("Could not find server URL in stdout: " + all.text);
  }

  return serverUrl;
}

async function checkBuildOutput(dir: string) {
  const distDir = path.join(dir, "dist");
  const files = await readdir(distDir);
  expect(files.some(f => f.endsWith(".js"))).toBe(true);
  expect(files.some(f => f.endsWith(".html"))).toBe(true);
  expect(files.some(f => f.endsWith(".css"))).toBe(true);
}

describe.each(["true", "false"])("development: %s", developmentString => {
  const development = developmentString === "true";
  const tempDirWithFiles = (name: string, files: Record<string, string>) =>
    tempDir(name + (development ? "-dev" : "-prod"), files);
  const normalizeHTML = normalizeHTMLFn(development);
  const env = {
    ...bunEnv,
    NODE_PORT: "0",
    NODE_ENV: development ? undefined : "production",
  };

  const devServerLabel = development ? " dev server" : "";
  describe("react spa (no tailwind)", async () => {
    let dir: string;
    beforeEach(async () => {
      dir = tempDirWithFiles("react-spa-no-tailwind", {
        "README.md": "Hello, world!",
      });

      await cp(path.join(__dirname, "react-spa-no-tailwind"), dir, {
        recursive: true,
        force: true,
      });
    });

    test.todoIf(isCI)("dev server", async () => {
      await using process = Bun.spawn([bunExe(), "create", "./index.jsx"], {
        cwd: dir,
        env: env,
        stdout: "pipe",
        stdin: "ignore",
      });
      const all = { text: "" };
      const serverUrl = await getServerUrl(process, all);

      try {
        const browser = await getPuppeteerBrowser();
        var page = await browser.newPage();
        await page.goto(serverUrl, { waitUntil: "networkidle0" });

        const content = await page.evaluate(() => document.documentElement.innerHTML);

        expect(normalizeHTML(content)).toMatchSnapshot();

        expect(
          all.text
            .replace(/v\d+\.\d+\.\d+(?:\s*\([a-f0-9]+\))?(?:-debug)?/g, "v*.*.*") // Handle version with git hash
            .replace(/\[\d+\.?\d*m?s\]/g, "[*ms]")
            .replace(/@\d+\.\d+\.\d+/g, "@*.*.*")
            .replace(/\d+\.\d+\s*ms/g, "*.** ms")
            .replace(/^\s+/gm, "") // Remove leading spaces
            .replace(/installed react(-dom)?@\d+\.\d+\.\d+/g, "installed react$1@*.*.*") // Handle react versions
            .trim()
            .replaceAll(serverUrl, "http://[SERVER_URL]"),
        ).toMatchSnapshot();
      } finally {
        process.kill();
        await Promise.resolve(page!?.close?.({ runBeforeUnload: false }));
      }
    });

    test.todoIf(isWindows)("build", async () => {
      {
        const process = Bun.spawn([bunExe(), "create", "./index.jsx"], {
          cwd: dir,
          env: env,
          stdout: "pipe",
          stdin: "ignore",
        });
        const all = { text: "" };
        const serverUrl = await getServerUrl(process, all);
        process.kill();
      }

      const process = Bun.spawn([bunExe(), "run", "build"], {
        cwd: dir,
        env: env,
        stdout: "pipe",
      });

      await process.exited;
      await checkBuildOutput(dir);
    });
  });

  describe("react spa (tailwind)", async () => {
    let dir: string;
    beforeEach(async () => {
      dir = tempDirWithFiles("react-spa-tailwind", {
        "index.tsx": await Bun.file(path.join(__dirname, "tailwind.tsx")).text(),
      });
    });

    test.todoIf(isCI)("dev server", async () => {
      const process = Bun.spawn([bunExe(), "create", "./index.tsx"], {
        cwd: dir,
        env: env,
        stdout: "pipe",
        stdin: "ignore",
      });
      const all = { text: "" };
      const serverUrl = await getServerUrl(process, all);
      console.log(serverUrl);

      try {
        var page = await (await getPuppeteerBrowser()).newPage();
        await page.goto(serverUrl, { waitUntil: "networkidle0" });

        // Check that React root exists and has Tailwind classes
        const root = await page.$("#root");
        expect(root).toBeTruthy();

        const content = await page.evaluate(() => document.documentElement.outerHTML);
        expect(normalizeHTML(content)).toMatchSnapshot();

        expect(
          all.text
            .replace(/v\d+\.\d+\.\d+(?:\s*\([a-f0-9]+\))?(?:-debug)?/g, "v*.*.*")
            .replace(/\[\d+\.?\d*m?s\]/g, "[*ms]")
            .replace(/@\d+\.\d+\.\d+/g, "@*.*.*")
            .replace(/\d+\.\d+\s*ms/g, "*.** ms")
            .replace(/^\s+/gm, "")
            .replace(/installed (react(-dom)?|tailwindcss)@\d+\.\d+\.\d+/g, "installed $1@*.*.*")
            .trim()
            .replaceAll(serverUrl, "http://[SERVER_URL]"),
        ).toMatchSnapshot();
      } finally {
        process.kill();
        await Promise.resolve(page!?.close?.({ runBeforeUnload: false }));
      }
    });

    test.todoIf(isWindows)("build", async () => {
      {
        const process = Bun.spawn([bunExe(), "create", "./index.tsx"], {
          cwd: dir,
          env: env,
          stdout: "pipe",
          stdin: "ignore",
        });
        const all = { text: "" };
        const serverUrl = await getServerUrl(process, all);
        process.kill();
      }

      const process = Bun.spawn([bunExe(), "run", "build"], {
        cwd: dir,
        env: env,
        stdout: "pipe",
      });

      await process.exited;
      await checkBuildOutput(dir);
    });
  });

  describe(
    "shadcn/ui",
    async () => {
      let dir: string;
      beforeEach(async () => {
        dir = tempDirWithFiles("shadcn-ui", {
          "index.tsx": await Bun.file(path.join(__dirname, "shadcn.tsx")).text(),
        });
      });

      test(
        "dev server",
        async () => {
          const process = Bun.spawn([bunExe(), "create", "./index.tsx"], {
            cwd: dir,
            env: env,
            stdout: "pipe",
            stdin: "ignore",
          });
          const all = { text: "" };
          const serverUrl = await getServerUrl(process, all);
          console.log(serverUrl);
          console.log(dir);
          try {
            var page = await (await getPuppeteerBrowser()).newPage();
            await page.goto(serverUrl, { waitUntil: "networkidle0" });

            // Check that React root exists and has Shadcn components
            const root = await page.$("#root");
            expect(root).toBeTruthy();

            const content = await page.evaluate(() => document.documentElement.innerHTML);
            expect(content).toContain("shadcn"); // Basic check for Shadcn classes

            // Check for components.json
            const componentsJson = await Bun.file(path.join(dir, "components.json")).exists();
            expect(componentsJson).toBe(true);

            expect(
              all.text
                .replace(/v\d+\.\d+\.\d+(?:\s*\([a-f0-9]+\))?(?:-debug)?/g, "v*.*.*")
                .replace(/\[\d+\.?\d*m?s\]/g, "[*ms]")
                .replace(/@\d+\.\d+\.\d+/g, "@*.*.*")
                .replace(/\d+\.\d+\s*ms/g, "*.** ms")
                .replace(/^\s+/gm, "")
                .replace(
                  /installed (react(-dom)?|@radix-ui\/.*|tailwindcss|class-variance-authority|clsx|lucide-react|tailwind-merge)@\d+\.\d+\.\d+/g,
                  "installed $1@*.*.*",
                )
                .trim()
                .replaceAll(serverUrl, "http://[SERVER_URL]"),
            ).toMatchSnapshot();
          } finally {
            process.kill();
            await Promise.resolve(page!?.close?.({ runBeforeUnload: false }));
          }
        },
        1000 * 100,
      );

      test.todoIf(isWindows)("build", async () => {
        {
          const process = Bun.spawn([bunExe(), "create", "./index.tsx"], {
            cwd: dir,
            env: env,
            stdout: "pipe",
            stdin: "ignore",
          });
          const all = { text: "" };
          const serverUrl = await getServerUrl(process, all);
          process.kill();
        }

        const process = Bun.spawn([bunExe(), "run", "build"], {
          cwd: dir,
          env: env,
          stdout: "pipe",
        });

        await process.exited;
        await checkBuildOutput(dir);
      });
    },
    1000 * 100,
  );
});

function normalizeHTMLFn(development: boolean = true) {
  return (html: string) =>
    html
      .split("\n")
      .map(line => {
        // First trim the line
        const trimmed = line.trim();
        if (!trimmed) return "";

        if (!development) {
          // Replace chunk hashes in stylesheet and script tags
          return trimmed.replace(
            /<(link rel="stylesheet" crossorigin="" href|script type="module" crossorigin="" src)="\/chunk-[a-zA-Z0-9]+\.(css|js)("><\/script>|">)/g,
            (_, tagStart, ext) => {
              if (ext === "css") {
                return `<${tagStart}="/chunk-[HASH].css">`;
              }
              return `<${tagStart}="/chunk-[HASH].js"></script>`;
            },
          );
        }

        // In development mode, replace non-deterministic generation IDs
        return trimmed
          .replace(/\/_bun\/client\/(.*?-[a-z0-9]{8})[a-z0-9]{8}\.js/, "/_bun/client/$1[NONDETERMINISTIC].js")
          .replace(/\/_bun\/asset\/[a-z0-9]{8}[a-z0-9]{8}\.[a-z]+/, "/_bun/asset/[ASSET_HASH].css");
      })
      .filter(Boolean)
      .join("\n")
      .trim();
}
