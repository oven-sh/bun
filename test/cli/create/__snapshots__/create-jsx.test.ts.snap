// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`development: true react spa (no tailwind) dev server 1`] = `
"<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>index | Powered by Bun</title>
<link rel="stylesheet" href="/_bun/asset/[ASSET_HASH].css"><link rel="stylesheet" href="/_bun/asset/ddeb04f41bc00a80.css"></head>
<body>
<div id="root"><div class="app"><main class="container"><div class="hero"><div class="logo animate-bounce">🥟</div><h1>Welcome to <span class="gradient-text">Bun</span></h1><p class="description">The all-in-one JavaScript runtime &amp; toolkit designed for speed</p><div class="cta-buttons"><a href="https://bun.sh" class="button primary" target="_blank" rel="noopener noreferrer">Get Started</a><a href="https://github.com/oven-sh/bun" class="button secondary" target="_blank" rel="noopener noreferrer">View on GitHub</a></div><div class="stats"><div class="stat"><span class="stat-value">3x</span><span class="stat-label">Bun Bun Bun</span></div><div class="stat"><span class="stat-value">0.5s</span><span class="stat-label">Average Install Time</span></div><div class="stat"><span class="stat-value">Extremely</span><span class="stat-label">Node.js Compatible</span></div></div></div><section class="features-section"><h2>Why Choose Bun?</h2><div class="features"><div class="feature hover:scale-105"><div class="feature-icon">⚡️</div><h3>Lightning Fast</h3><p>Built from scratch in <span class="highlight">Zig</span>, Bun is focused on performance and developer experience</p></div><div class="feature hover:scale-105"><div class="feature-icon">🎯</div><h3>All-in-One</h3><p>Bundler, test runner, and npm-compatible package manager in a single tool</p></div><div class="feature hover:scale-105"><div class="feature-icon">🚀</div><h3>JavaScript Runtime</h3><p>Drop-in replacement for Node.js with <span class="highlight">3x faster</span> startup time</p></div><div class="feature hover:scale-105"><div class="feature-icon">📦</div><h3>Package Management</h3><p>Native package manager that can install dependencies up to <span class="highlight">30x faster</span> than npm</p></div><div class="feature hover:scale-105"><div class="feature-icon">🧪</div><h3>Testing Made Simple</h3><p>Built-in test runner with Jest-compatible API and snapshot testing</p></div><div class="feature hover:scale-105"><div class="feature-icon">🔥</div><h3>Hot Reloading</h3><p>Lightning-fast hot module replacement (HMR) for rapid development</p></div></div></section></main><footer class="footer"><div class="footer-content"><div class="footer-logo"><span class="logo-small">🥟</span><span class="footer-text">Built with Bun</span></div><nav class="footer-links"><a href="https://bun.sh/docs" class="footer-link hover:text-accent" target="_blank" rel="noopener noreferrer">Documentation</a><a href="https://github.com/oven-sh/bun" class="footer-link hover:text-accent" target="_blank" rel="noopener noreferrer">GitHub</a><a href="https://bun.sh/discord" class="footer-link hover:text-accent" target="_blank" rel="noopener noreferrer">Discord</a><a href="https://bun.sh/blog" class="footer-link hover:text-accent" target="_blank" rel="noopener noreferrer">Blog</a></nav></div></footer></div></div>
<script type="module" crossorigin="" src="/_bun/client/index-00000000[NONDETERMINISTIC].js"></script>
</body>"
`;

exports[`development: true react spa (no tailwind) dev server 2`] = `
"create  index.build.ts     build
create  index.css          css
create  index.html         html
create  index.client.tsx   bun
create  package.json       npm
📦 Auto-installing 3 detected dependencies
$ bun --only-missing install classnames react-dom@19 react@19
bun add v*.*.*
installed classnames@*.*.*
installed react-dom@*.*.*
installed react@*.*.*
4 packages installed [*ms]
--------------------------------
✨ React project configured
Development - frontend dev server with hot reload
bun dev
Production - build optimized assets
bun run build
Happy bunning! 🐇
Bun v*.*.* dev server ready in *.** ms
url: http://[SERVER_URL]/"
`;

exports[`development: true react spa (tailwind) dev server 1`] = `
"<html><head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>index | Powered by Bun</title>
<link rel="stylesheet" href="/_bun/asset/[ASSET_HASH].css"></head>
<body>
<div id="root"><div class="min-h-screen bg-gradient-to-b from-gray-900 to-black text-white"><div class="max-w-6xl mx-auto px-4 py-20"><div class="text-center mb-16"><h1 class="text-6xl font-bold mb-6"><span class="text-purple-400">bun create</span> for React</h1><p class="text-xl text-gray-300 mb-8">Start a React dev server instantly from a single component file</p><div class="bg-gray-800 p-4 rounded-lg flex items-center justify-between max-w-lg mx-auto mb-8"><code class="text-purple-400">bun create ./MyComponent.tsx</code><button class="bg-purple-600 hover:bg-purple-700 px-4 py-2 rounded transition">Copy</button></div></div><div class="grid md:grid-cols-3 gap-8 mb-20"><div class="bg-gray-800 p-6 rounded-lg"><h3 class="text-xl font-semibold mb-4">Zero Config</h3><p class="text-gray-300">Just write your React component and run. No setup needed.</p></div><div class="bg-gray-800 p-6 rounded-lg"><h3 class="text-xl font-semibold mb-4">Auto Dependencies</h3><p class="text-gray-300">Automatically detects and installs required npm packages.</p></div><div class="bg-gray-800 p-6 rounded-lg"><h3 class="text-xl font-semibold mb-4">Tool Detection</h3><p class="text-gray-300">Recognizes Tailwind, animations, and UI libraries automatically.</p></div></div><div class="bg-gray-800 rounded-lg p-8 mb-20"><h2 class="text-3xl font-bold mb-6">How it Works</h2><div class="space-y-4"><div class="flex items-start gap-4"><div class="bg-purple-600 rounded-full p-2 mt-1">1</div><div><h3 class="font-semibold mb-2">Create Component</h3><p class="text-gray-300">Write your React component in a .tsx file</p></div></div><div class="flex items-start gap-4"><div class="bg-purple-600 rounded-full p-2 mt-1">2</div><div><h3 class="font-semibold mb-2">Run Command</h3><p class="text-gray-300">Execute bun create with your file path</p></div></div><div class="flex items-start gap-4"><div class="bg-purple-600 rounded-full p-2 mt-1">3</div><div><h3 class="font-semibold mb-2">Start Developing</h3><p class="text-gray-300">Dev server starts instantly with hot reload</p></div></div></div></div><div class="text-center"><h2 class="text-3xl font-bold mb-6">Ready to Try?</h2><div class="space-x-4"><a href="https://bun.sh/docs" class="inline-block bg-purple-600 hover:bg-purple-700 px-6 py-3 rounded-lg transition">Read Docs</a><a href="https://github.com/oven-sh/bun" class="inline-block bg-gray-700 hover:bg-gray-600 px-6 py-3 rounded-lg transition">GitHub →</a></div></div></div></div></div>
<script type="module" crossorigin="" src="/_bun/client/index-00000000[NONDETERMINISTIC].js"></script>
</body></html>"
`;

exports[`development: true react spa (tailwind) dev server 2`] = `
"create  index.build.ts     build
create  index.css          css
create  index.html         html
create  index.client.tsx   bun
create  bunfig.toml        bun
create  package.json       npm
📦 Auto-installing 4 detected dependencies
$ bun --only-missing install tailwindcss bun-plugin-tailwind react-dom@19 react@19
bun add v*.*.*
installed tailwindcss@*.*.*
installed bun-plugin-tailwind@*.*.*
installed react-dom@*.*.*
installed react@*.*.*
7 packages installed [*ms]
--------------------------------
✨ React + Tailwind project configured
Development - frontend dev server with hot reload
bun dev
Production - build optimized assets
bun run build
Happy bunning! 🐇
Bun v*.*.* dev server ready in *.** ms
url: http://[SERVER_URL]/"
`;

exports[`development: true shadcn/ui dev server 1`] = `
"create  lib/utils.ts         shadcn
create  src/index.css        shadcn
create  index.build.ts       bun
create  index.client.tsx     bun
create  index.css            css
create  index.html           html
create  styles/globals.css   shadcn
create  bunfig.toml          bun
create  package.json         npm
create  tsconfig.json        tsc
create  components.json      shadcn
📦 Auto-installing 9 detected dependencies
$ bun --only-missing install lucide-react tailwindcss bun-plugin-tailwind tailwindcss-animate class-variance-authority clsx tailwind-merge react@^18 react-dom@^18
bun add v*.*.*
installed lucide-react@*.*.*
installed tailwindcss@*.*.*
installed bun-plugin-tailwind@*.*.*
installed tailwindcss-animate@*.*.*
installed class-variance-authority@*.*.*
installed clsx@*.*.*
installed tailwind-merge@*.*.*
installed react@*.*.*
installed react-dom@*.*.*
14 packages installed [*ms]
😎 Setting up shadcn/ui components
$ bun x shadcn@canary add -y button badge card
- components/ui/button.tsx
- components/ui/badge.tsx
- components/ui/card.tsx
--------------------------------
✨ React + shadcn/ui + Tailwind project configured
Development - frontend dev server with hot reload
bun dev
Production - build optimized assets
bun run build
Happy bunning! 🐇
Bun v*.*.* dev server ready in *.** ms
url: http://[SERVER_URL]/"
`;

exports[`development: false react spa (no tailwind) dev server 1`] = `
"<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>index | Powered by Bun</title>
<link rel="icon" type="image/x-icon" href="https://bun.sh/favicon.ico">
<link rel="stylesheet" crossorigin="" href="/chunk-[HASH].css"><script type="module" crossorigin="" src="/chunk-[HASH].js"></script></head>
<body>
<div id="root"><div class="app"><main class="container"><div class="hero"><div class="logo animate-bounce">🥟</div><h1>Welcome to <span class="gradient-text">Bun</span></h1><p class="description">The all-in-one JavaScript runtime &amp; toolkit designed for speed</p><div class="cta-buttons"><a href="https://bun.sh" class="button primary" target="_blank" rel="noopener noreferrer">Get Started</a><a href="https://github.com/oven-sh/bun" class="button secondary" target="_blank" rel="noopener noreferrer">View on GitHub</a></div><div class="stats"><div class="stat"><span class="stat-value">3x</span><span class="stat-label">Bun Bun Bun</span></div><div class="stat"><span class="stat-value">0.5s</span><span class="stat-label">Average Install Time</span></div><div class="stat"><span class="stat-value">Extremely</span><span class="stat-label">Node.js Compatible</span></div></div></div><section class="features-section"><h2>Why Choose Bun?</h2><div class="features"><div class="feature hover:scale-105"><div class="feature-icon">⚡️</div><h3>Lightning Fast</h3><p>Built from scratch in <span class="highlight">Zig</span>, Bun is focused on performance and developer experience</p></div><div class="feature hover:scale-105"><div class="feature-icon">🎯</div><h3>All-in-One</h3><p>Bundler, test runner, and npm-compatible package manager in a single tool</p></div><div class="feature hover:scale-105"><div class="feature-icon">🚀</div><h3>JavaScript Runtime</h3><p>Drop-in replacement for Node.js with <span class="highlight">3x faster</span> startup time</p></div><div class="feature hover:scale-105"><div class="feature-icon">📦</div><h3>Package Management</h3><p>Native package manager that can install dependencies up to <span class="highlight">30x faster</span> than npm</p></div><div class="feature hover:scale-105"><div class="feature-icon">🧪</div><h3>Testing Made Simple</h3><p>Built-in test runner with Jest-compatible API and snapshot testing</p></div><div class="feature hover:scale-105"><div class="feature-icon">🔥</div><h3>Hot Reloading</h3><p>Lightning-fast hot module replacement (HMR) for rapid development</p></div></div></section></main><footer class="footer"><div class="footer-content"><div class="footer-logo"><span class="logo-small">🥟</span><span class="footer-text">Built with Bun</span></div><nav class="footer-links"><a href="https://bun.sh/docs" class="footer-link hover:text-accent" target="_blank" rel="noopener noreferrer">Documentation</a><a href="https://github.com/oven-sh/bun" class="footer-link hover:text-accent" target="_blank" rel="noopener noreferrer">GitHub</a><a href="https://bun.sh/discord" class="footer-link hover:text-accent" target="_blank" rel="noopener noreferrer">Discord</a><a href="https://bun.sh/blog" class="footer-link hover:text-accent" target="_blank" rel="noopener noreferrer">Blog</a></nav></div></footer></div></div>
</body>"
`;

exports[`development: false react spa (no tailwind) dev server 2`] = `
"create  index.build.ts     build
create  index.css          css
create  index.html         html
create  index.client.tsx   bun
create  package.json       npm
📦 Auto-installing 3 detected dependencies
$ bun --only-missing install classnames react-dom@19 react@19
bun add v*.*.*
installed classnames@*.*.*
installed react-dom@*.*.*
installed react@*.*.*
4 packages installed [*ms]
--------------------------------
✨ React project configured
Development - frontend dev server with hot reload
bun dev
Production - build optimized assets
bun run build
Happy bunning! 🐇
Bun v*.*.* ready in *.** ms
url: http://[SERVER_URL]/"
`;

exports[`development: false react spa (tailwind) dev server 1`] = `
"<html><head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>index | Powered by Bun</title>
<link rel="icon" type="image/x-icon" href="https://bun.sh/favicon.ico">
<link rel="stylesheet" crossorigin="" href="/chunk-[HASH].css"><script type="module" crossorigin="" src="/chunk-[HASH].js"></script></head>
<body>
<div id="root"><div class="min-h-screen bg-gradient-to-b from-gray-900 to-black text-white"><div class="max-w-6xl mx-auto px-4 py-20"><div class="text-center mb-16"><h1 class="text-6xl font-bold mb-6"><span class="text-purple-400">bun create</span> for React</h1><p class="text-xl text-gray-300 mb-8">Start a React dev server instantly from a single component file</p><div class="bg-gray-800 p-4 rounded-lg flex items-center justify-between max-w-lg mx-auto mb-8"><code class="text-purple-400">bun create ./MyComponent.tsx</code><button class="bg-purple-600 hover:bg-purple-700 px-4 py-2 rounded transition">Copy</button></div></div><div class="grid md:grid-cols-3 gap-8 mb-20"><div class="bg-gray-800 p-6 rounded-lg"><h3 class="text-xl font-semibold mb-4">Zero Config</h3><p class="text-gray-300">Just write your React component and run. No setup needed.</p></div><div class="bg-gray-800 p-6 rounded-lg"><h3 class="text-xl font-semibold mb-4">Auto Dependencies</h3><p class="text-gray-300">Automatically detects and installs required npm packages.</p></div><div class="bg-gray-800 p-6 rounded-lg"><h3 class="text-xl font-semibold mb-4">Tool Detection</h3><p class="text-gray-300">Recognizes Tailwind, animations, and UI libraries automatically.</p></div></div><div class="bg-gray-800 rounded-lg p-8 mb-20"><h2 class="text-3xl font-bold mb-6">How it Works</h2><div class="space-y-4"><div class="flex items-start gap-4"><div class="bg-purple-600 rounded-full p-2 mt-1">1</div><div><h3 class="font-semibold mb-2">Create Component</h3><p class="text-gray-300">Write your React component in a .tsx file</p></div></div><div class="flex items-start gap-4"><div class="bg-purple-600 rounded-full p-2 mt-1">2</div><div><h3 class="font-semibold mb-2">Run Command</h3><p class="text-gray-300">Execute bun create with your file path</p></div></div><div class="flex items-start gap-4"><div class="bg-purple-600 rounded-full p-2 mt-1">3</div><div><h3 class="font-semibold mb-2">Start Developing</h3><p class="text-gray-300">Dev server starts instantly with hot reload</p></div></div></div></div><div class="text-center"><h2 class="text-3xl font-bold mb-6">Ready to Try?</h2><div class="space-x-4"><a href="https://bun.sh/docs" class="inline-block bg-purple-600 hover:bg-purple-700 px-6 py-3 rounded-lg transition">Read Docs</a><a href="https://github.com/oven-sh/bun" class="inline-block bg-gray-700 hover:bg-gray-600 px-6 py-3 rounded-lg transition">GitHub →</a></div></div></div></div></div>
</body></html>"
`;

exports[`development: false react spa (tailwind) dev server 2`] = `
"create  index.build.ts     build
create  index.css          css
create  index.html         html
create  index.client.tsx   bun
create  bunfig.toml        bun
create  package.json       npm
📦 Auto-installing 4 detected dependencies
$ bun --only-missing install tailwindcss bun-plugin-tailwind react-dom@19 react@19
bun add v*.*.*
installed tailwindcss@*.*.*
installed bun-plugin-tailwind@*.*.*
installed react-dom@*.*.*
installed react@*.*.*
7 packages installed [*ms]
--------------------------------
✨ React + Tailwind project configured
Development - frontend dev server with hot reload
bun dev
Production - build optimized assets
bun run build
Happy bunning! 🐇
Bun v*.*.* ready in *.** ms
url: http://[SERVER_URL]/"
`;

exports[`development: false shadcn/ui dev server 1`] = `
"create  lib/utils.ts         shadcn
create  src/index.css        shadcn
create  index.build.ts       bun
create  index.client.tsx     bun
create  index.css            css
create  index.html           html
create  styles/globals.css   shadcn
create  bunfig.toml          bun
create  package.json         npm
create  tsconfig.json        tsc
create  components.json      shadcn
📦 Auto-installing 9 detected dependencies
$ bun --only-missing install lucide-react tailwindcss bun-plugin-tailwind tailwindcss-animate class-variance-authority clsx tailwind-merge react@^18 react-dom@^18
bun add v*.*.*
installed lucide-react@*.*.*
installed tailwindcss@*.*.*
installed bun-plugin-tailwind@*.*.*
installed tailwindcss-animate@*.*.*
installed class-variance-authority@*.*.*
installed clsx@*.*.*
installed tailwind-merge@*.*.*
installed react@*.*.*
installed react-dom@*.*.*
14 packages installed [*ms]
😎 Setting up shadcn/ui components
$ bun x shadcn@canary add -y button badge card
- components/ui/button.tsx
- components/ui/badge.tsx
- components/ui/card.tsx
--------------------------------
✨ React + shadcn/ui + Tailwind project configured
Development - frontend dev server with hot reload
bun dev
Production - build optimized assets
bun run build
Happy bunning! 🐇
Bun v*.*.* ready in *.** ms
url: http://[SERVER_URL]/"
`;
